### ISSUE TITLE
WMV files are downloaded as ASX files in Firefox

### ISSUE THREAD

**Issue by NateWr** [pkp/pkp-lib#7474]
**Describe the bug**
When downloading a WMV file such as `example.wmv`, the Firefox download prompt will name the file`example.asf`. This does not happen in Chrome, where the file's name is shown as `example.wmv`. 

I've figured out what is happening but I'm not sure how to address it.

When `example.wmv` is uploaded, the mimetype is being recorded as `video/x-ms-asf` in the `files` table. This is a (correct) parent mimetype for `video/x-ms-wmv`. See: https://stackoverflow.com/questions/53672122/why-wmv-files-have-mime-type-video-x-ms-asf-instead-of-video-x-ms-wmv.

When the file is downloaded, the filename and mimetype are passed in the response headers:

https://github.com/pkp/pkp-lib/blob/main/classes/services/PKPFileService.inc.php#L161-L167

The headers will have a `Content-Type` of `video/x-ms-asf` and the `filename` in the `Content-Disposition` header will be `example.wmv`.

It seems that Firefox is using the `Content-Type` to override the supplied filename.

**To Reproduce**
1. Get a valid WMV file (not just a text file renamed example.wmv)
1. Go to a submission and upload a WMV file to a submission file area (eg - Submission Files, Files for Review, Revisions).
3. Try to download that file in Firefox.
4. See the download filename is set to `.asf`

**What application are you using?**
3.3.0-x

**Additional information**
The following file, [example-wmv.zip](https://github.com/pkp/pkp-lib/files/7602194/example-wmv.zip), contains a `.wmv` file that can be used to reproduce this. However, this file should not be used for any purposes other than testing this bug.

**Comment by jonasraoni** [comment]
There's some extra details here: https://bugzilla.mozilla.org/show_bug.cgi?id=1688306

Looks like it's not going to be changed 😁
So we'll have to ensure the mime-type is a perfect fit for its extension or perhaps send an extensionless content-type `application/octet-stream` if the action is related to a download button 🤔

**Comment by NateWr** [comment]
Good find @jonasraoni! I've added a comment there because it seems to me like Firefox has already written some exceptions to the rule and this is a good candidate for an exception, since `video/x-ms-asf` is a valid mimetype for `.wma` and `.wmv` files.

> send an extensionless content-type

I think we'll have the same problem here, but in Chrome as well, because according to a comment on that mozilla bug report Chrome uses the mimetype in cases where no file extension is passed.

**Comment by henriqueramos** [comment]
We could create some filter to change the mimetype response when detected a call from Firefox.

Here it's the [mime type list](https://searchfox.org/mozilla-central/source/third_party/rust/mime_guess/src/mime_types.rs#86) used by Firefox to choose a file extension.

If you send the mime type as`video/x-ms-wmv` the file extension will be the correct one, as see [here](https://searchfox.org/mozilla-central/source/third_party/rust/mime_guess/src/mime_types.rs#1364).

**Comment by jonasraoni** [comment]
> I think we'll have the same problem here, but in Chrome as well,

What problem? If we send `Content-type: application/octet-stream` for all files (that are not going to be `$inline`d), I think it should work ok, even though it's not beautiful 🤔

**Comment by NateWr** [comment]
> We could create some filter to change the mimetype response when detected a call from Firefox.

I try to avoid browser sniffing if at all possible because it hasn't historically been very reliable. Handling this mimetype specially on our end may be the only way to fix this, but I'd much rather leave it to Firefox to handle. Exceptions like this lead to a cluttered and stale codebase over time.

> If we send Content-type: application/octet-stream for all files (that are not going to be $inlined), I think it should work ok

I would be nervous about delivering an inaccurate mime type header. Although we do `inline` in some cases, in others we don't, and the browser may still wish to display this file (eg - PDFs, images, videos, rendered JSON/CSV). If we make everything `application/octet-stream`, we are effectively overriding any behaviour specified by the browser or the user's settings. In these cases, I don't like doing things our own way because we can't reasonably anticipate the consequences of our actions.

**Comment by NateWr** [comment]
Oh, I've already received a reply on the Mozilla bug, and it looks like this may be  a more recent regression that will be fixed: https://bugzilla.mozilla.org/show_bug.cgi?id=1734606. :crossed_fingers:

**Comment by NateWr** [comment]
@ctgraham reminded us that we already have some mimetype mapping in place:

https://github.com/pkp/pkp-lib/blob/da52e6aa2973d0395674257aacc74effab8584d3/classes/core/PKPString.inc.php#L394-L414

Extending this with the following might do the trick:

```
'wmv:video/x-ms-asf' => 'video/x-ms-wmv',
'wma:video/x-ms-asf' => 'audio/x-ms-wma',
```

I'd like to see if there is any movement on Firefox's end before committing to this customized mimetype mapping though.

**Comment by NateWr** [comment]
Another example of a similar issue with CSV files downloaded as `.txt`: https://forum.pkp.sfu.ca/t/csv-galleys-converted-to-txt/70966/3

**Comment by NateWr** [comment]
I may be reading this wrong, but it looks like both tickets on Firefox's tracker are marked as `wontfix`. I think we should expect to make our own workaround as suggested in [this comment](https://github.com/pkp/pkp-lib/issues/7474#issuecomment-977995884).

It looks like we may need one for CSV files too:

```
'csv:text/plain' => 'text/csv',
```

**Comment by heike-r** [comment]
If I may add another problematic file type: docm is wrongly downloaded as docx by Firefox in OJS 3.3.0.8
We want to use docm as internediate file type in our production workflow. I can upload it as submission file but when clicking on the file to download, it is recognized as docx by Firefox. Same es above, it works fine in Chrome and also if I use the 'download all files' button in Firefox.

**Comment by NateWr** [comment]
Thanks everyone! The following PRs fix mimetype detection for wma, wmv, csv and docm files.

PRs:
https://github.com/pkp/pkp-lib/pull/7632 (`stable-3_3_0`)
https://github.com/pkp/pkp-lib/pull/7633 (`main`)

Tests:
https://github.com/pkp/ojs/pull/3281 (`stable_3_3_0`)

This will fix mimetype detection for newly uploaded files. Files that were already uploaded will still have the old mimetype. These will be fixed when upgrading to 3.4.0.

If you want to fix the mimetype for files that were already uploaded and you don't want to wait until you upgrade to 3.4.0, you can run the following SQL queries. These have not been widely tested. Please make a backup of your database before running them.

```sql
update files set mimetype='video/x-ms-wma' where mimetype='video/x-ms-asf' and path LIKE '%.wma';
update files set mimetype='video/x-ms-wmv' where mimetype='video/x-ms-asf' and path LIKE '%.wmv';
update files set mimetype='application/vnd.ms-word.document.macroEnabled.12' where mimetype='application/vnd.openxmlformats-officedocument.wordprocessingml.document' and path LIKE '%.docm';
update files set mimetype='text/csv' where mimetype='text/plain' and path LIKE '%.csv';
```

**Comment by NateWr** [comment]
Merged to `stable-3_3_0` and `main`.
